### PR TITLE
reduce semantic boost when matched docs are few

### PR DESCRIPTION
**Pull request by yingjieMiao**
@ymatsuda

### CODE REVIEW

File: kifi-backend/modules/search/app/com/keepit/search/query/TextQuery.scala [hunk 1]
```diff
@@ -223,7 +224,12 @@ class TextScorer(weight: TextWeight, personalScorer: Scorer, regularScorer: Scor
   private[this] var docR = if (regularScorer == null) NO_MORE_DOCS else -1
   private[this] var scoredDoc = -1
   private[this] var scoreVal = 0.0f
-  private[this] val semanticScoreBase = (1.0f - semanticBoost)
+  private[this] val adjustedSemanticBoost = {
+    val good = if (semanticScorer != null) semanticScorer.getChildren().exists(scorer => scorer.child.asInstanceOf[SemanticVectorScorer].hasGoodQuality) else false
+    if (good) semanticBoost else 0.1f
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/query/TextQuery.scala]: Can we have a smooth transition? Let's avoid a sudden behavior change.

Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/query/TextQuery.scala]: did you mean 0.1f is too small? We should reduce this little by little?

Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/query/TextQuery.scala]: 0.1f is not too small. What I don't like is a drastic change around the threshold. I think it should change from 0.0f to semanticBoost according to the number of URIs used for sv construction.

Comment (yingjieMiao) [inline kifi-backend/modules/search/app/com/keepit/search/query/TextQuery.scala]: I see. Will change this.

**Comment by yingjieMiao** [comment]
@ymatsuda 

OK to merge?

File: kifi-backend/modules/search/app/com/keepit/search/query/TextQuery.scala [hunk 2]
```diff
@@ -223,7 +224,16 @@ class TextScorer(weight: TextWeight, personalScorer: Scorer, regularScorer: Scor
   private[this] var docR = if (regularScorer == null) NO_MORE_DOCS else -1
   private[this] var scoredDoc = -1
   private[this] var scoreVal = 0.0f
-  private[this] val semanticScoreBase = (1.0f - semanticBoost)
+  private[this] val adjustedSemanticBoost = {
+    val n = if (semanticScorer != null) {
+      semanticScorer.getChildren().map(scorer => scorer.child.asInstanceOf[SemanticVectorScorer].getNumPayloadsUsed).foldLeft(0)(_ max _)
+    } else 0
+
+    val adjust = 1.0/(1 + pow(1.5, 5 - n))
```
Comment (ymatsuda) [inline kifi-backend/modules/search/app/com/keepit/search/query/TextQuery.scala:232]: nice!

**Comment by ymatsuda** [comment]
lgtm!

**Pull request merged**